File: lab5-collections/haskell/tttColl.hs
module Main where

import qualified Data.Map as Map
import qualified Data.List as List

data Player 
    = Player_X | Player_O
    deriving (Eq)

instance (Show Player) where
    show Player_X = "X"
    show Player_O = "O"

data Cell
    = X | O | B
    deriving (Eq, Ord)
instance (Show Cell) where
    show X = "X"
    show O = "O"
    show B = "."

data Board
    = Board { board_rows :: [[Cell]] } -- list of lists (ie rows) of cells
    deriving (Eq, Ord)

instance (Show Board) where
    show (Board rows) =
        -- show each row and put the rows on separate lines:
        joinLines $ map showRow $ rows
            where
            showRow :: [Cell] -> String
            showRow cells =
                -- concatenate cells
                concat $ map show cells

changeCell (Board rows) cellValue (row,col) =
    Board newRows
        where
        newRows = 
            updateListElem 
                row -- update this index
                (updateListElem col (const cellValue)) 
                        -- using this function, which
                        -- updates the cell in the row
                rows -- in this list

-- modifying one element in a list is a pain...
-- one needs to construct a new list from scratch
updateListElem index fn list = 
    (take index list) -- elements up to index: no change
    ++    -- concatenate
    [fn (list !! index)] -- the element that is changed
    ++    -- concatenate
    (drop (index + 1) list) -- the rest does not change

board1 = 
    Board
    [[X,X,O],
     [O,B,X],
     [O,B,B]]

board2 = 
    changeCell board1 X (1,1)

data Event
    = EventMove
      {
        ev_player :: Player,
        ev_pos :: (Int, Int) -- a pair (tuple) type
      }
    | EventPause
      {
        ev_player :: Player
      }
    deriving (Eq)

isPause (EventPause _) = True
isPause _ = False

instance (Show Event) where
    show (EventMove player (row, col)) =
        "Move of player " ++ show player ++ 
            " to (" ++ show row ++ "," ++ show col ++ ")"
    show (EventPause player) =
        "Pause of player " ++ show player
        

events = 
    [EventMove Player_X (0,0),
     EventPause Player_X,
     EventMove Player_O (1,0)]

players =
    [ ev_player event | event <- events ]

events2 =
    [ event | event <- events, not (isPause event) ]

board2score =
    Map.fromList [(board1, 2.0), (board2, 0.0)]

nonBlankFields =
    [cell | row <- board_rows board1, cell <- row, cell /= B]
    
main =
    do
    putStrLn $ "board1 = \n" ++ show board1
    putStrLn $ "board2 = \n" ++ show board2
    putStrLn $ "events = \n" ++ (showListIndented events) -- one per line
    putStrLn $ "players = " ++ (show players)
    putStrLn $ "events2 = \n" ++ (showListIndented events2)
    putStrLn $ "nonBlankFields = " ++ (show nonBlankFields)
    case Map.lookup board1 board2score of
        Nothing -> return ()
        Just score -> putStrLn $ "board1 score = " ++ show score
    where
    showListIndented list =
        joinLines $ map indent $ map show list
        where
        indent line = "    " ++ line

joinLines = List.intercalate "\n"

File: lab5-collections/python/tttColl.py
# -*- coding: utf-8 -*-
class Board:
    def __init__(self, rows): # constructor
        self.rows = rows
        self.check()
        
    def __repr__(self):  # formatting (like toString())
        result = ""
        for row in self.rows:
            for cell in row:
                result = result + cell
            result = result + "\n"
        return result
            
    def check(self):
        # do some dynamic type checking:
        assert len(self.rows) == 3
        for row in self.rows:
            assert len(row) == 3
            for cell in row:
                assert cell in ['X','O','B']
                
    def changeCell(self, cellValue, row, col):
        self.rows[row][col] = cellValue

board1 = Board(
    [['X','X','O'],
     ['X','B','O'],
     ['O','B','B']])
    # Python has no native arrays - use lists instead
    # Python has no enumerated types - use strings 'X' etc.

board2 = Board(
    [['X','X','O'],
     ['X','X','O'],
     ['O','O','X']])

print(board1)
board1.changeCell('X',1,1)
print(board1)

class Move:
    def __init__(self,player,pos): # constructor
        self.player = player
        self.pos = pos
    def __repr__(self): # formatting (like toString())
        return \
            "A move of player %s at position (%s,%s)" % \
             (self.player, self.pos[0], self.pos[1])
                # Python version of printf
    # backslash at end of line means: continue on next line

class Pause:
    def __init__(self,player): # constructor
        self.player = player
    def __repr__(self): # formatting (like toString())
        return \
            "A pause of player %s" % (self.player)
                # Python version of printf


# a heterogeneous list
events = [Move("X",[0,0]), Pause("X"), Move("O",[1,0])]

move = Move("X",[0,0])

# print all elements from the list
for event in events:
    print(event)

# empty line
print()

# search for an element of class Pause and print it
for event in events:
    if event.__class__ == Pause:
        print(event)
        break # stop the for loop

# empty line
print()

# compute players
players = [event.player for event in events]

for player in players:
    print(player)

# empty line
print()

# remove all pauses
events2 = \
    [event for event in events if event.__class__ != Pause]
    
for event in events2:
    print(event)

# non-blank fields from a board:
nonBlankFields = [cell for row in board1.rows for cell in row if cell != "B"]
print("nonBlankFields = ", nonBlankFields)

# heterogeneous dictionary:
board2score = { board1 : 2.0, board2 : "dunno" }
board2score[ board2 ] = 0.0 # like array update
print(board2score[ board1 ]) # like array lookup

